feat: erc20 support

kincaidoneil opened the pull request:
**ERC-20s**
- Open, fund and close payment channels denominated in any ERC-20
- Send and receive paychan claims denominated in an ERC-20
- Initial channel open also requires an approve transaction, so approve Machinomy contract to transfer unlimited tokens from the configured Ethereum account
  - Secure since account must explicitly open/deposit for Machinomy contract to transfer tokens to itself
  - Overestimate combined gas limit for both txs (no way to combine them into 1 tx, or estimate the gas for the second until the first is executed)
- Update claim validation to ensure the correct token contract (or the absence of one if used with ether)
- Support tokens with varying decimal places (convert from asset scale 9)
- Auto load token symbol and scale from the contract address
- Test using Ganache

**Other changes**
- docs(readme): better overview, add new config options
- fix: after channel funding/closing, use events to confirm state updates (likely safer than fetching by `channelId` to update state)
- feat: option for custom Machinomy contract address
- feat: option for custom Ethers wallet instance

kincaidoneil committed:
feat: erc20 support

ERC-20s
- open, fund and close payment channels denominated in any ERC-20
- send and receive paychan claims denominated in an ERC-20
- initial channel open also requires an approve transaction
- approve Machinomy contract to transfer unlimited tokens from the configured Ethereum account
- secure since account must explicitly open/deposit for Machinomy contract to transfer tokens to itself
- overestimate combined gas limit for both txs (no way to combine them into 1 tx, or estimate the gas for the second until the first is executed)
- update claim validation to ensure the correct token contract (or the absence of one if used with ether)
- support tokens with varying decimal places (convert from asset scale 9)
- auto load token symbol and scale from the contract address

Other features
- fix: after channel funding/closing, use events to confirm state updates (likely safer than fetching by channelId to update state)
- feat: option for custom Machinomy contract address
- feat: option for custom Ethers wallet instance

kincaidoneil committed:
test(erc20): send and receive tokens with ganache

- tests mirror eth tests, but using ganache rather than live testnet
- add ganache as devDep
- add OpenZeppelin contract artifacts (for ERC-20s) as dep

kincaidoneil committed:
docs(readme): new config options, edits

- rename "tokenContract" to "tokenAddress"

kincaidoneil committed:
fix: miscellaneous cleanup

- readme clarifications
- allow custom Ethers providers to not break existing API
- check for sufficient ether/token balance before submitting opens and deposits
- fix minor todos
- add serialized paychan types
- add more function signatures
- add sanity check if channel closed between a state and a later state

kincaidoneil committed:
feat: add mainnet token contract

karzak commented on README.md:
```diff
@@ -50,29 +73,40 @@ This plugin uses an asset scale of 9 and units of _gwei_, and **not** 18, or wei
+>
+> Servers _do_ automatically open channels. If a client has opened a channel with a value above the configurable `minIncomingChannelAmount`, the server will automatically open a channel back to the client with a value of `outgoingChannelAmount`. When the channel is half empty, the server will also automatically top up the value of the channel by the `outgoingChannelAmount`.
+>
+> The balance configuration has been simplified for servers. Clients must prefund before sending any packets through a server, and if a client fulfills packets sent to them through a server, the server will automatically settle such that they owe 0 to the client. This configuration was chosen because it assumes the server would extend no credit to anonymous client, and clients extend little credi ...
```
typo: to anonymous client -> to an anonymous client

karzak commented on src/__tests__/send-receive-token.test.ts:
```diff
@@ -0,0 +1,141 @@
+  )
+  const tokenContract = await tokenFactory.deploy()
+
+  // Mint test tokens for both accounts (100 units of the token, assuming of -18 base)
```
typo:  assuming of -> assuming

karzak commented:
nit: in `account.ts` we still reference 'units of wei' (https://github.com/interledgerjs/ilp-plugin-ethereum/blob/53bc585ed26d70aa3b855826536b82a703a4e1f2/src/account.ts#L293, https://github.com/interledgerjs/ilp-plugin-ethereum/blob/53bc585ed26d70aa3b855826536b82a703a4e1f2/src/account.ts#L339, https://github.com/interledgerjs/ilp-plugin-ethereum/blob/53bc585ed26d70aa3b855826536b82a703a4e1f2/src/account.ts#L485) in places were it now makes more sense to say 'base units'.

kincaidoneil committed:
docs: address comments

kincaidoneil committed:
fix: bug loading ERC-20 metadata

- Fix error when loading ERC-20 asset scale
- DAI incorrectly implements ERC-20 spec, so add special case for its symbol
- When comparing two cached channel states, use shared util to check if channel closed in between
- Add comments explaining why the state transition is safe while checking deposit
